### ISSUE TITLE
Crossgen2 perfmap format version support

### ISSUE THREAD

**Issue by trylek** [dotnet/sdk#18813]
As of

https://github.com/dotnet/runtime/pull/53792

Crossgen2 newly supports a new command-line option <code>--perfmap-format-version</code> that currently has two legal values, 0 and 1, where 0 corresponds to the "old style Crossgen1" perfmap naming format and 1 is the "new style Crossgen2" naming format. In particular, historically Crossgen1 used to name perfmap files like this:

<code>&lt;assembly-name&gt;.ni.{&lt;MVID of the MSIL file&gt;}.map</code>

This is insufficient and cumbersome in Crossgen2 composite mode where multiple assemblies get compiled into a single native executable; moreover the pre-existing scheme was insufficient anyway as symbol indexation needs to take targeting OS and architecture into consideration and that wasn't reflected in the <code>MVID</code>. The new naming in version 1 is like this:

<code>&lt;assembly-name&gt;.ni.r2rmap</code>

The current POR is that we drop the <code>{MVID}</code> bit and change <code>.map</code> to <code>.r2rmap</code> to avoid clashes with traditional .map files; this will need adjusting the following SDK source files:

https://github.com/dotnet/sdk/blob/main/src/Tasks/Microsoft.NET.Build.Tasks/ResolveReadyToRunCompilers.cs
https://github.com/dotnet/sdk/blob/main/src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs
https://github.com/dotnet/sdk/blob/main/src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs

We also need to adjust the script calling them:

https://github.com/dotnet/sdk/blob/main/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets

These changes should let SDK make Crossgen2 produce the new perfmap format version 1 and ultimately switch over installer build in the runtime repo to produce the new perfmap style files. According to my understanding the bulk of the follow-up work will involve consuming the metadata in the perfmap file for the purpose of symbol indexation; I believe that should be tracked by a separate work item.

This change may need a counterpart perfview change and possibly additional downstream changes I'm not yet aware of; these will probably merit creating additional specific work items as we discover them.

Thanks

Tomas

/cc @dotnet/crossgen-contrib
@dotnet/dotnet-diag 
@tommcdon 
@brianrob 
@hoyosjs 
@mikem8361

**Comment by marcpopMSFT** [comment]
@tommcdon to triage and assign if this is needed in net 6.

**Comment by davmason** [comment]
Tom is out on vacation until August. The perfview changes can happen out of band, we need to know and track any runtime work that happens because we have about a month left.

@trylek @hoyosjs @mikem8361 - how much work is necessary to be in the runtime to fully support the scenario?

**Comment by trylek** [comment]
I believe we need basically two changes:

1) An SDK change tweaking PrepareReadyToRunCompilation et al and the Microsoft.NET.Publish.targets to accept a new property optionally defining the perfmap version format. SDK should be already consuming Crossgen2 Preview 7 that supports this option. When downlevel targeting .NET 5, it must assume the old naming scheme.

2) In the runtime repo, there are two places that need adjusting: S.P.Corelib build in the coreclr partition and framework crossgenning in the installer partition. These must be conscious changes correlated with the downstream symbol indexation logic; moreover they are blocked on consuming the updated SDK supporting the new property.

Based on these initial assumptions I believe the two changes will first have a chance to combine upon the RC fork in August even though earlier testing should be possible in the SDK repo that receives regular runtime updates.

Thanks

Tomas

**Comment by davmason** [comment]
Thanks Tomas! Do we have the appropriate work items tracking this? Tom is out and I'm generally unaware of this work item. It doesn't sound like any work from the diagnostics team is needed for 6, but will be out of band in perfview. Is that right?

**Comment by trylek** [comment]
Well, this was originally intended to be "the work item" tracking the outstanding work but I can easily create two separate ones for the SDK and runtime parts. I can easily make the two fixes in accordance with Tom's suggestion in our related e-mail thread on the issue but it would be great it @hoyosjs and @brianrob could confirm I'm not missing something in the overall plan. I assume that the "work from the diagnostics team" will largely involve setting up the perfmap symbol indexation mechanism but I think it's a completely new feature not necessarily tied to the .NET 6 release.

**Comment by hoyosjs** [comment]
The symuploader work needs to happen + pipe it through arcade and then flow it up.

**Comment by trylek** [comment]
I have merged in the SDK change adding support for perfmap format version. @hoyosjs / @brianrob, any suggestions regarding the next step? Should we file a new issue to track the symbol upload / indexation logic in arcade? I'm afraid that the remaining work is pretty arcane for me, it would be great if someone more familiar with this logic could pick it up (apart from the fact that I'm OOF for one more week next week).

**Comment by hoyosjs** [comment]
You can put it there as there's some work that needs to be done there to the SFX framework too (The globbing needs to be updated). Then runtime and WindowsDesktop use that to crossgen the shared framework I believe. ASP will need to happen separately as they don't use that. Also @trylek, do you know if runtime uses the live crossgen2, or does it use the one in the SDK? Because if so this becomes post-RC1 work.

**Comment by hoyosjs** [comment]
Added the details of the missing work items to consume this. Except for the last one, I thing all are needed if we are going to get this for .NET 6.
